### PR TITLE
Add support to copy multiple messages

### LINKED ISSUE #3748: Copy multiple messages
Feature request: Would be nice to be able to copy multiple messages at a time.
screen
So for example if I wanted to copy the above two messages to send to someone else that could be highly convenient. The copy icon goes away when you select multiple messages, leaves you with only the ability to delete.

I realize there's some difficulty with how to perform the actual copying. I would assume the "right thing to do" would be to grab all of the text selected and mash it into one big string that gets put in paste clipboard.

### PR DESCRIPTION
**Pull request by f41c0r**
Closes #3748.
Single copy still works as before, but now so does multiple copy:
anim
anim_mult

### CODE REVIEW

**Comment by mammuth** [comment]
Nice, thank you! 
I was about to implement the same, since I started missing it.

**Comment by f41c0r** [comment]
Thanks for the heads up, added code to sort based on messageRecieved time

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 1]
```diff
@@ -192,15 +194,34 @@ public void run() {
...
+    }
+    String result = bodyBuilder.toString();
+    if (result.equals("")) return;
+    clipboard.setText(result.substring(0,(result.length()-1)));
+  }
 
-    ClipboardManager clipboard = (ClipboardManager)getActivity()
-        .getSystemService(Context.CLIPBOARD_SERVICE);
-    clipboard.setText(body);
+  private static class messageChronologicalOrderComparator implements Comparator<MessageRecord> {
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: classes always begin with uppercase

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 2]
```diff
@@ -144,7 +146,7 @@ private void setCorrectMenuVisibility(Menu menu) {
 
     if (messageRecords.size() > 1) {
       menu.findItem(R.id.menu_context_forward).setVisible(false);
-      menu.findItem(R.id.menu_context_copy).setVisible(false);
+      menu.findItem(R.id.menu_context_copy).setVisible(true);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: you can just remove this line

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 3]
```diff
@@ -192,15 +194,34 @@ public void run() {
...
+  private void handleCopyMessage(final List<MessageRecord> messageRecords) {
+    messageChronologicalOrderComparator mc = new messageChronologicalOrderComparator();
+    Collections.sort(messageRecords,mc);
+    StringBuilder bodyBuilder = new StringBuilder();
+    String body;
+    ClipboardManager clipboard = (ClipboardManager) getActivity()
+            .getSystemService(Context.CLIPBOARD_SERVICE);
+    for (MessageRecord message : messageRecords) {
+      body = message.getDisplayBody().toString();
+      if (body == null) continue;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: if not null

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 4]
```diff
@@ -192,15 +194,34 @@ public void run() {
...
   }
 
-  private void handleCopyMessage(MessageRecord message) {
-    String body = message.getDisplayBody().toString();
-    if (body == null) return;
+  private void handleCopyMessage(final List<MessageRecord> messageRecords) {
+    messageChronologicalOrderComparator mc = new messageChronologicalOrderComparator();
+    Collections.sort(messageRecords,mc);
+    StringBuilder bodyBuilder = new StringBuilder();
+    String body;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: why is this out of scope

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 5]
```diff
@@ -192,15 +194,34 @@ public void run() {
...
-    ClipboardManager clipboard = (ClipboardManager)getActivity()
-        .getSystemService(Context.CLIPBOARD_SERVICE);
-    clipboard.setText(body);
+  private static class messageChronologicalOrderComparator implements Comparator<MessageRecord> {
+    @Override
+    public int compare(MessageRecord lhs, MessageRecord rhs) {
+      if (lhs.getDateReceived() < rhs.getDateReceived()) return -1;
+      else if (lhs.getDateReceived() == rhs.getDateReceived()) return 0;
+      else return 1;
+    }
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: this class doesn't belong in here, can also probably be anonymous

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 6]
```diff
@@ -192,15 +194,34 @@ public void run() {
...
+    ClipboardManager clipboard = (ClipboardManager) getActivity()
+            .getSystemService(Context.CLIPBOARD_SERVICE);
+    for (MessageRecord message : messageRecords) {
+      body = message.getDisplayBody().toString();
+      if (body == null) continue;
+      bodyBuilder.append(body);
+      bodyBuilder.append('\n');
+    }
+    String result = bodyBuilder.toString();
+    if (result.equals("")) return;
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: if !textutils.isempty

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 7]
```diff
@@ -192,15 +194,34 @@ public void run() {
...
+    clipboard.setText(result.substring(0,(result.length()-1)));
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: the substring is gross

Comment (f41c0r) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: Well, I could do 
import org.apache.commons.lang3.StringUtils;
clipboard.setText(StringUtils.removeEnd(result,'\n'));

Definitely more readable, but then that adds another dependency, just to chop the end of a string...

**Comment by f41c0r** [comment]
Okay, cleaned up the code. About the ugliness of the result.substring:

I could do
import org.apache.commons.lang3.StringUtils;
clipboard.setText(StringUtils.removeEnd(result,'\n'));

Definitely more readable, but then that adds another dependency, just to chop the end of a string...

Or I could write a method with a name like "chopEnd" to do that, if you think that's more readable.

Please let me know if/what I need to do!

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 8]
```diff
@@ -192,13 +194,31 @@ public void run() {
...
+      @Override
+      public int compare(MessageRecord lhs, MessageRecord rhs) {
+        if (lhs.getDateReceived() < rhs.getDateReceived())
+          return -1;
+        else if (lhs.getDateReceived() == rhs.getDateReceived())
+          return 0;
+        else
+          return 1;
+      }
+    });
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: whitespace

File: src/org/thoughtcrime/securesms/ConversationFragment.java [hunk 9]
```diff
@@ -192,13 +194,31 @@ public void run() {
...
+    StringBuilder bodyBuilder = new StringBuilder();
+    ClipboardManager clipboard = (ClipboardManager) getActivity()
+            .getSystemService(Context.CLIPBOARD_SERVICE);
```
Comment (moxie0) [inline src/org/thoughtcrime/securesms/ConversationFragment.java]: align

**Comment by moxie0** [comment]
don't append the newline to the last record

**Comment by f41c0r** [comment]
Not sure what you meant by cleaning up whitespace, did I fix that correctly?

**Comment by McLoo** [comment]
@f41c0r guess that would be an empty line between 205 and 206

**Comment by f41c0r** [comment]
ait added a newline, and rebased back into one commit

**Comment by agrajaghh** [comment]
I think he was talking about `bodyBuilder.append('\n');` in line 215

**Comment by f41c0r** [comment]
what about line 215?

**Comment by agrajaghh** [comment]
@f41c0r 
sry, guess I read the code too quickly, you don't append a newline there for the last record. Don't know what moxie meant..

**Comment by moxie0** [comment]
in 2.25.0, thanks

in the future please make sure your commit messages are wrapped appropriately (first line < 76 chars, subsequent lines wrapped at 8)

**Pull request closed without merging**